### PR TITLE
Reduce memory footprint for Entity type

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

**Comment by AviAvni** [comment]
@dotnet-bot Test Ubuntu16.04 Release_default Build please

**Comment by AviAvni** [comment]
@dotnet-bot Test Ubuntu16.04 Release_default Build please

**Comment by AviAvni** [comment]
@dotnet-bot Test Ubuntu16.04 Release_fcs Build please

**Comment by AviAvni** [comment]
@dsyme 2 field left
``` entity_tycon_tcaug: TyconAugmentation      // None for types from .NET``` - this is not optional what to do on the TypeContents property in case of entity_opt_data is None?
```entity_typars: LazyWithContext<Typars, range>        // usually empty, might be hard``` - I'll try to work on this

**Comment by AviAvni** [comment]
@dotnet-bot Test Ubuntu16.04 Release_default Build please

File: src/fsharp/tast.fs [hunk 1]
```diff
@@ -574,12 +577,30 @@ type Entity =
       /// Used during codegen to hold the ILX representation indicating how to access the type 
       // MUTABILITY: only for unpickle linkage and caching
       mutable entity_il_repr_cache : CompiledTypeRepr cache
+
+      mutable entity_opt_data : EntityOptionalData option
     }
+
+    static member EmptyEntityOptData = { entity_compiled_name = None; entity_other_range = None; entity_kind = TyparKind.Type; entity_xmldoc = XmlDoc.Empty; entity_xmldocsig = ""; entity_tycon_abbrev = None; entity_tycon_repr_accessibility = TAccess []; entity_accessiblity = TAccess []; entity_exn_info = TExnNone }
```
Comment (dsyme) [inline src/fsharp/tast.fs]: As mentioned on previous PR, using a static property can be misleading when generating a mutable data structure.  Perhaps a static method ``MyEmptyEntityOptionalData()``

**Comment by AviAvni** [comment]
@dsyme running the statistic on the FSharp.Compiler.Private.fsproj 
```
~~~ Entity statistics: WithOptData 7743, WithoutOptData 5632
```

**Comment by AviAvni** [comment]
if my calculation is correct then:
image
the first one is for Entity
the the second one is for Val

**Comment by dsyme** [comment]
> ~~~ Entity statistics: WithOptData 7743, WithoutOptData 5632

Hmmm, I wonder if this is tuned right.  

1. Could you check the stats on the compilation of a simple file *when referencing all the standard references for .NET Framework or .NET Standard* when opening a bunch of namespaces and just a little code to trigger a search for a type name amongst all available types:

```fsharp
   open System
   open System.Text
   open System.Collections.Generic
   open System.Globalization
   open System.Net
   open System.Web
   ... // whatever else you want 

    let f1 () = new DateTime()
    let f2 () = Action(fun () -> ())
```

2. Is there a way to get the "cause" of the `WithoutOptData` cases?  Which field forced the creation of the optional data?

**Comment by AviAvni** [comment]
@dsyme second issue result https://github.com/AviAvni/visualfsharp/compare/fcs-memory-2...AviAvni:fcs-memory-2-measurment
```
~~~ Entity statistics: WithOptData 7743, WithoutOptData 5632
~~~ Entity statistics: entity_compiled_name: 48
~~~ Entity statistics: entity_other_range: 3896
~~~ Entity statistics: entity_kind: 106
~~~ Entity statistics: entity_xmldoc: 1848
~~~ Entity statistics: entity_xmldocsig: 2451
~~~ Entity statistics: entity_tycon_abbrev: 737
~~~ Entity statistics: entity_tycon_repr_accessibility: 3253
~~~ Entity statistics: entity_accessiblity: 3253
~~~ Entity statistics: entity_exn_info: 502
```
first part working on it

**Comment by AviAvni** [comment]
@dsyme 

and for 
```
open System
open System.IO
open System.IO.IsolatedStorage
open System.Security
open System.Security.AccessControl
open System.Security.Cryptography
open System.Security.Cryptography.X509Certificates
open System.Security.Permissions
open System.Security.Claims
open System.Security.Principal
open System.Security.Policy
open System.Resources
open System.Globalization
open System.Diagnostics
open System.Diagnostics.SymbolStore
open System.Diagnostics.Contracts
open System.Diagnostics.Tracing
open System.Collections
open System.Collections.Generic
open System.Threading
open System.Threading.Tasks
open System.Reflection
open System.Runtime.ConstrainedExecution
open System.Runtime.Serialization
open System.Runtime.Serialization.Formatters
open System.Runtime.Serialization.Formatters.Binary
open System.Runtime.ExceptionServices
open System.Runtime.Remoting
open System.Runtime.Remoting.Metadata
open System.Runtime.Remoting.Metadata.W3cXsd2001
open System.Runtime.Remoting.Proxies
open System.Runtime.Remoting.Services
open System.Runtime.Remoting.Contexts
open System.Runtime.Remoting.Lifetime
open System.Runtime.Remoting.Channels
open System.Runtime.Remoting.Messaging
open System.Runtime.CompilerServices
open System.Runtime.InteropServices
open System.Runtime.InteropServices.WindowsRuntime
open System.Runtime.Hosting
open System.Text
let AggregateException = AggregateException()
let Object = Object()
let Exception = Exception()
let SystemException = SystemException()
let OutOfMemoryException = OutOfMemoryException()
let StackOverflowException = StackOverflowException()
let DataMisalignedException = DataMisalignedException()
let ExecutionEngineException = ExecutionEngineException()
let MemberAccessException = MemberAccessException()
let AccessViolationException = AccessViolationException()
let ApplicationException = ApplicationException()
let AppDomainManager = AppDomainManager()
let AppDomainSetup = AppDomainSetup()
let AppDomainUnloadedException = AppDomainUnloadedException()
let ArgumentException = ArgumentException()
let ArgumentNullException = ArgumentNullException()
let ArgumentOutOfRangeException = ArgumentOutOfRangeException()
let ArithmeticException = ArithmeticException()
let ArrayTypeMismatchException = ArrayTypeMismatchException()
let BadImageFormatException = BadImageFormatException()
let CannotUnloadAppDomainException = CannotUnloadAppDomainException()
let TypeUnloadedException = TypeUnloadedException()
let ContextMarshalException = ContextMarshalException()
let ContextStaticAttribute = ContextStaticAttribute()
let DivideByZeroException = DivideByZeroException()
let DuplicateWaitObjectException = DuplicateWaitObjectException()
let EntryPointNotFoundException = EntryPointNotFoundException()
let DllNotFoundException = DllNotFoundException()
let EventArgs = EventArgs()
let FieldAccessException = FieldAccessException()
let FlagsAttribute = FlagsAttribute()
let FormatException = FormatException()
let IndexOutOfRangeException = IndexOutOfRangeException()
let InsufficientMemoryException = InsufficientMemoryException()
let InsufficientExecutionStackException = InsufficientExecutionStackException()
let InvalidCastException = InvalidCastException()
let InvalidOperationException = InvalidOperationException()
let InvalidProgramException = InvalidProgramException()
let InvalidTimeZoneException = InvalidTimeZoneException()
let MethodAccessException = MethodAccessException()
let MissingFieldException = MissingFieldException()
let MissingMemberException = MissingMemberException()
let MissingMethodException = MissingMethodException()
let MulticastNotSupportedException = MulticastNotSupportedException()
let NonSerializedAttribute = NonSerializedAttribute()
let NotFiniteNumberException = NotFiniteNumberException()
let NotImplementedException = NotImplementedException()
let NotSupportedException = NotSupportedException()
let NullReferenceException = NullReferenceException()
let ObsoleteAttribute = ObsoleteAttribute()
let OperationCanceledException = OperationCanceledException()
let OverflowException = OverflowException()
let ParamArrayAttribute = ParamArrayAttribute()
let PlatformNotSupportedException = PlatformNotSupportedException()
let Random = Random()
let RankException = RankException()
let SerializableAttribute = SerializableAttribute()
let STAThreadAttribute = STAThreadAttribute()
let MTAThreadAttribute = MTAThreadAttribute()
let TimeoutException = TimeoutException()
let TimeZoneNotFoundException = TimeZoneNotFoundException()
let TypeAccessException = TypeAccessException()
let TypeLoadException = TypeLoadException()
let UnauthorizedAccessException = UnauthorizedAccessException()
let Version = Version()
let ThreadStaticAttribute = ThreadStaticAttribute()
let DirectoryNotFoundException = DirectoryNotFoundException()
let DriveNotFoundException = DriveNotFoundException()
let EndOfStreamException = EndOfStreamException()
let FileLoadException = FileLoadException()
let FileNotFoundException = FileNotFoundException()
let IOException = IOException()
let MemoryStream = MemoryStream()
let PathTooLongException = PathTooLongException()
let StringWriter = StringWriter()
let IsolatedStorageException = IsolatedStorageException()
let XmlSyntaxException = XmlSyntaxException()
let SuppressUnmanagedCodeSecurityAttribute = SuppressUnmanagedCodeSecurityAttribute()
let UnverifiableCodeAttribute = UnverifiableCodeAttribute()
let AllowPartiallyTrustedCallersAttribute = AllowPartiallyTrustedCallersAttribute()
let SecurityCriticalAttribute = SecurityCriticalAttribute()
let SecurityTreatAsSafeAttribute = SecurityTreatAsSafeAttribute()
let SecuritySafeCriticalAttribute = SecuritySafeCriticalAttribute()
let SecurityTransparentAttribute = SecurityTransparentAttribute()
let HostSecurityManager = HostSecurityManager()
let SecureString = SecureString()
let SecurityException = SecurityException()
let HostProtectionException = HostProtectionException()
let VerificationException = VerificationException()
let CryptoKeySecurity = CryptoKeySecurity()
let EventWaitHandleSecurity = EventWaitHandleSecurity()
let FileSecurity = FileSecurity()
let DirectorySecurity = DirectorySecurity()
let MutexSecurity = MutexSecurity()
let PrivilegeNotHeldException = PrivilegeNotHeldException()
let RegistrySecurity = RegistrySecurity()
let AuthorizationRuleCollection = AuthorizationRuleCollection()
let CryptographicException = CryptographicException()
let CryptographicUnexpectedOperationException = CryptographicUnexpectedOperationException()
let RNGCryptoServiceProvider = RNGCryptoServiceProvider()
let ToBase64Transform = ToBase64Transform()
let FromBase64Transform = FromBase64Transform()
let CspParameters = CspParameters()
let CryptoConfig = CryptoConfig()
let DESCryptoServiceProvider = DESCryptoServiceProvider()
let DSACryptoServiceProvider = DSACryptoServiceProvider()
let DSASignatureDeformatter = DSASignatureDeformatter()
let DSASignatureFormatter = DSASignatureFormatter()
let HMACMD5 = HMACMD5()
let HMACRIPEMD160 = HMACRIPEMD160()
let HMACSHA1 = HMACSHA1()
let HMACSHA256 = HMACSHA256()
let HMACSHA384 = HMACSHA384()
let HMACSHA512 = HMACSHA512()
let MACTripleDES = MACTripleDES()
let MD5CryptoServiceProvider = MD5CryptoServiceProvider()
let PKCS1MaskGenerationMethod = PKCS1MaskGenerationMethod()
let RC2CryptoServiceProvider = RC2CryptoServiceProvider()
let RIPEMD160Managed = RIPEMD160Managed()
let RSACryptoServiceProvider = RSACryptoServiceProvider()
let RSAOAEPKeyExchangeDeformatter = RSAOAEPKeyExchangeDeformatter()
let RSAOAEPKeyExchangeFormatter = RSAOAEPKeyExchangeFormatter()
let RSAPKCS1KeyExchangeDeformatter = RSAPKCS1KeyExchangeDeformatter()
let RSAPKCS1KeyExchangeFormatter = RSAPKCS1KeyExchangeFormatter()
let RSAPKCS1SignatureDeformatter = RSAPKCS1SignatureDeformatter()
let RSAPKCS1SignatureFormatter = RSAPKCS1SignatureFormatter()
let RijndaelManaged = RijndaelManaged()
let SHA1CryptoServiceProvider = SHA1CryptoServiceProvider()
let SHA1Managed = SHA1Managed()
let SHA256Managed = SHA256Managed()
let SHA384Managed = SHA384Managed()
let SHA512Managed = SHA512Managed()
let SignatureDescription = SignatureDescription()
let TripleDESCryptoServiceProvider = TripleDESCryptoServiceProvider()
let X509Certificate = X509Certificate()
let HostProtectionAttribute = HostProtectionAttribute()
let GacIdentityPermission = GacIdentityPermission()
let ClaimsIdentity = ClaimsIdentity()
let ClaimsPrincipal = ClaimsPrincipal()
let IdentityReferenceCollection = IdentityReferenceCollection()
let IdentityNotMappedException = IdentityNotMappedException()
let AllMembershipCondition = AllMembershipCondition()
let ApplicationDirectoryMembershipCondition = ApplicationDirectoryMembershipCondition()
let ApplicationTrust = ApplicationTrust()
let Evidence = Evidence()
let TrustManagerContext = TrustManagerContext()
let PolicyException = PolicyException()
let GacInstalled = GacInstalled()
let GacMembershipCondition = GacMembershipCondition()
let MissingManifestResourceException = MissingManifestResourceException()
let MissingSatelliteAssemblyException = MissingSatelliteAssemblyException()
let CultureNotFoundException = CultureNotFoundException()
let DateTimeFormatInfo = DateTimeFormatInfo()
let GregorianCalendar = GregorianCalendar()
let HebrewCalendar = HebrewCalendar()
let HijriCalendar = HijriCalendar()
let UmAlQuraCalendar = UmAlQuraCalendar()
let ChineseLunisolarCalendar = ChineseLunisolarCalendar()
let JapaneseLunisolarCalendar = JapaneseLunisolarCalendar()
let JulianCalendar = JulianCalendar()
let KoreanLunisolarCalendar = KoreanLunisolarCalendar()
let PersianCalendar = PersianCalendar()
let TaiwanLunisolarCalendar = TaiwanLunisolarCalendar()
let IdnMapping = IdnMapping()
let JapaneseCalendar = JapaneseCalendar()
let KoreanCalendar = KoreanCalendar()
let StringInfo = StringInfo()
let TaiwanCalendar = TaiwanCalendar()
let ThaiBuddhistCalendar = ThaiBuddhistCalendar()
let NumberFormatInfo = NumberFormatInfo()
let DebuggerStepThroughAttribute = DebuggerStepThroughAttribute()
let DebuggerStepperBoundaryAttribute = DebuggerStepperBoundaryAttribute()
let DebuggerHiddenAttribute = DebuggerHiddenAttribute()
let DebuggerNonUserCodeAttribute = DebuggerNonUserCodeAttribute()
let StackTrace = StackTrace()
let StackFrame = StackFrame()
let SymDocumentType = SymDocumentType()
let SymLanguageType = SymLanguageType()
let SymLanguageVendor = SymLanguageVendor()
let PureAttribute = PureAttribute()
let ContractInvariantMethodAttribute = ContractInvariantMethodAttribute()
let ContractReferenceAssemblyAttribute = ContractReferenceAssemblyAttribute()
let ContractRuntimeIgnoredAttribute = ContractRuntimeIgnoredAttribute()
let ContractArgumentValidatorAttribute = ContractArgumentValidatorAttribute()
let ContractAbbreviatorAttribute = ContractAbbreviatorAttribute()
let EventSourceAttribute = EventSourceAttribute()
let NonEventAttribute = NonEventAttribute()
let EventSourceException = EventSourceException()
let EventDataAttribute = EventDataAttribute()
let EventFieldAttribute = EventFieldAttribute()
let EventIgnoreAttribute = EventIgnoreAttribute()
let CaseInsensitiveComparer = CaseInsensitiveComparer()
let CaseInsensitiveHashCodeProvider = CaseInsensitiveHashCodeProvider()
let Queue = Queue()
let ArrayList = ArrayList()
let Stack = Stack()
let Hashtable = Hashtable()
let SortedList = SortedList()
let KeyNotFoundException = KeyNotFoundException()
let AbandonedMutexException = AbandonedMutexException()
let SynchronizationContext = SynchronizationContext()
let HostExecutionContext = HostExecutionContext()
let HostExecutionContextManager = HostExecutionContextManager()
let LockRecursionException = LockRecursionException()
let Mutex = Mutex()
let Overlapped = Overlapped()
let ReaderWriterLock = ReaderWriterLock()
let SemaphoreFullException = SemaphoreFullException()
let SynchronizationLockException = SynchronizationLockException()
let ThreadInterruptedException = ThreadInterruptedException()
let ThreadStateException = ThreadStateException()
let WaitHandleCannotBeOpenedException = WaitHandleCannotBeOpenedException()
let ManualResetEventSlim = ManualResetEventSlim()
let CancellationTokenSource = CancellationTokenSource()
let ParallelOptions = ParallelOptions()
let TaskCanceledException = TaskCanceledException()
let TaskSchedulerException = TaskSchedulerException()
let TaskFactory = TaskFactory()
let ConcurrentExclusiveSchedulerPair = ConcurrentExclusiveSchedulerPair()
let AmbiguousMatchException = AmbiguousMatchException()
let AssemblyName = AssemblyName()
let AssemblyNameProxy = AssemblyNameProxy()
let CustomAttributeFormatException = CustomAttributeFormatException()
let InvalidFilterCriteriaException = InvalidFilterCriteriaException()
let ObfuscationAttribute = ObfuscationAttribute()
let TargetException = TargetException()
let TargetParameterCountException = TargetParameterCountException()
let PrePrepareMethodAttribute = PrePrepareMethodAttribute()
let FormatterConverter = FormatterConverter()
let OptionalFieldAttribute = OptionalFieldAttribute()
let OnSerializingAttribute = OnSerializingAttribute()
let OnSerializedAttribute = OnSerializedAttribute()
let OnDeserializingAttribute = OnDeserializingAttribute()
let OnDeserializedAttribute = OnDeserializedAttribute()
let SerializationException = SerializationException()
let ObjectIDGenerator = ObjectIDGenerator()
let SurrogateSelector = SurrogateSelector()
let InternalRM = InternalRM()
let SoapMessage = SoapMessage()
let SoapFault = SoapFault()
let BinaryFormatter = BinaryFormatter()
let HandleProcessCorruptedStateExceptionsAttribute = HandleProcessCorruptedStateExceptionsAttribute()
let ObjRef = ObjRef()
let RemotingException = RemotingException()
let ServerException = ServerException()
let RemotingTimeoutException = RemotingTimeoutException()
let InternalRemotingServices = InternalRemotingServices()
let SoapTypeAttribute = SoapTypeAttribute()
let SoapMethodAttribute = SoapMethodAttribute()
let SoapFieldAttribute = SoapFieldAttribute()
let SoapParameterAttribute = SoapParameterAttribute()
let SoapAttribute = SoapAttribute()
let SoapDateTime = SoapDateTime()
let SoapDuration = SoapDuration()
let SoapTime = SoapTime()
let SoapDate = SoapDate()
let SoapYearMonth = SoapYearMonth()
let SoapYear = SoapYear()
let SoapMonthDay = SoapMonthDay()
let SoapDay = SoapDay()
let SoapMonth = SoapMonth()
let SoapHexBinary = SoapHexBinary()
let SoapBase64Binary = SoapBase64Binary()
let SoapInteger = SoapInteger()
let SoapPositiveInteger = SoapPositiveInteger()
let SoapNonPositiveInteger = SoapNonPositiveInteger()
let SoapNonNegativeInteger = SoapNonNegativeInteger()
let SoapNegativeInteger = SoapNegativeInteger()
let SoapAnyUri = SoapAnyUri()
let SoapQName = SoapQName()
let SoapNotation = SoapNotation()
let SoapNormalizedString = SoapNormalizedString()
let SoapToken = SoapToken()
let SoapLanguage = SoapLanguage()
let SoapName = SoapName()
let SoapIdrefs = SoapIdrefs()
let SoapEntities = SoapEntities()
let SoapNmtoken = SoapNmtoken()
let SoapNmtokens = SoapNmtokens()
let SoapNcName = SoapNcName()
let SoapId = SoapId()
let SoapIdref = SoapIdref()
let SoapEntity = SoapEntity()
let ProxyAttribute = ProxyAttribute()
let EnterpriseServicesHelper = EnterpriseServicesHelper()
let TrackingServices = TrackingServices()
let Context = Context()
let SynchronizationAttribute = SynchronizationAttribute()
let ClientSponsor = ClientSponsor()
let ClientChannelSinkStack = ClientChannelSinkStack()
let ServerChannelSinkStack = ServerChannelSinkStack()
let TransportHeaders = TransportHeaders()
let OneWayAttribute = OneWayAttribute()
let RemotingSurrogateSelector = RemotingSurrogateSelector()
let StringFreezingAttribute = StringFreezingAttribute()
let CallConvCdecl = CallConvCdecl()
let CallConvStdcall = CallConvStdcall()
let CallConvThiscall = CallConvThiscall()
let CallConvFastcall = CallConvFastcall()
let CompilerGeneratedAttribute = CompilerGeneratedAttribute()
let DiscardableAttribute = DiscardableAttribute()
let DisablePrivateReflectionAttribute = DisablePrivateReflectionAttribute()
let CompilerGlobalScopeAttribute = CompilerGlobalScopeAttribute()
let ExtensionAttribute = ExtensionAttribute()
let MethodImplAttribute = MethodImplAttribute()
let FixedAddressValueTypeAttribute = FixedAddressValueTypeAttribute()
let UnsafeValueTypeAttribute = UnsafeValueTypeAttribute()
let HasCopySemanticsAttribute = HasCopySemanticsAttribute()
let ScopelessEnumAttribute = ScopelessEnumAttribute()
let SpecialNameAttribute = SpecialNameAttribute()
let SuppressIldasmAttribute = SuppressIldasmAttribute()
let NativeCppClassAttribute = NativeCppClassAttribute()
let ReferenceAssemblyAttribute = ReferenceAssemblyAttribute()
let RuntimeCompatibilityAttribute = RuntimeCompatibilityAttribute()
let CallerFilePathAttribute = CallerFilePathAttribute()
let CallerLineNumberAttribute = CallerLineNumberAttribute()
let CallerMemberNameAttribute = CallerMemberNameAttribute()
let IDispatchConstantAttribute = IDispatchConstantAttribute()
let IUnknownConstantAttribute = IUnknownConstantAttribute()
let TypeIdentifierAttribute = TypeIdentifierAttribute()
let AllowReversePInvokeCallsAttribute = AllowReversePInvokeCallsAttribute()
let ComRegisterFunctionAttribute = ComRegisterFunctionAttribute()
let ComUnregisterFunctionAttribute = ComUnregisterFunctionAttribute()
let ComConversionLossAttribute = ComConversionLossAttribute()
let ComImportAttribute = ComImportAttribute()
let PreserveSigAttribute = PreserveSigAttribute()
let InAttribute = InAttribute()
let OutAttribute = OutAttribute()
let OptionalAttribute = OptionalAttribute()
let SetWin32ContextInIDispatchAttribute = SetWin32ContextInIDispatchAttribute()
let COMException = COMException()
let ExternalException = ExternalException()
let InvalidOleVariantTypeException = InvalidOleVariantTypeException()
let MarshalDirectiveException = MarshalDirectiveException()
let SEHException = SEHException()
let InvalidComObjectException = InvalidComObjectException()
let RegistrationServices = RegistrationServices()
let SafeArrayRankMismatchException = SafeArrayRankMismatchException()
let SafeArrayTypeMismatchException = SafeArrayTypeMismatchException()
let TypeLibConverter = TypeLibConverter()
let ReadOnlyArrayAttribute = ReadOnlyArrayAttribute()
let WriteOnlyArrayAttribute = WriteOnlyArrayAttribute()
let ApplicationActivator = ApplicationActivator()
let StringBuilder = StringBuilder()
let ASCIIEncoding = ASCIIEncoding()
let DecoderExceptionFallback = DecoderExceptionFallback()
let DecoderExceptionFallbackBuffer = DecoderExceptionFallbackBuffer()
let DecoderFallbackException = DecoderFallbackException()
let DecoderReplacementFallback = DecoderReplacementFallback()
let EncoderExceptionFallback = EncoderExceptionFallback()
let EncoderExceptionFallbackBuffer = EncoderExceptionFallbackBuffer()
let EncoderFallbackException = EncoderFallbackException()
let EncoderReplacementFallback = EncoderReplacementFallback()
let UnicodeEncoding = UnicodeEncoding()
let UTF7Encoding = UTF7Encoding()
let UTF8Encoding = UTF8Encoding()
let UTF32Encoding = UTF32Encoding()
```

the result are

```
~~~ Entity statistics: WithOptData 636, WithoutOptData 4926
~~~ Entity statistics: entity_compiled_name: 48
~~~ Entity statistics: entity_other_range: 0
~~~ Entity statistics: entity_kind: 106
~~~ Entity statistics: entity_xmldoc: 0
~~~ Entity statistics: entity_xmldocsig: 612
~~~ Entity statistics: entity_tycon_abbrev: 178
~~~ Entity statistics: entity_tycon_repr_accessibility: 46
~~~ Entity statistics: entity_accessiblity: 46
~~~ Entity statistics: entity_exn_info: 2
```

**Comment by dsyme** [comment]
@AviAvni fixed the comment

I took a look and there is only one place where we assume XmlDocSig is populated, in this dodgy code below.  There are better ways to check for the `unit` type, e.g. `isUnitTy` .

```
                if arg.Type.HasTypeDefinition && arg.Type.TypeDefinition.XmlDocSig = "T:Microsoft.FSharp.Core.unit" then "()" 
```

If we adjust that line then I think we should try ignoring the XmlDocSig when unpickling the signature data in TastPickle.fs and instead always initially populate entity_xmldocsig as if it is None.  You can also always write None when pickling (existing F# binaries will have the field populated).

**Comment by AviAvni** [comment]
@dsyme I ignored the XmlDocSig in TastPickle
and isUnitType not accept FSharpType so I used CompiledName

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs [hunk 2]
```diff
@@ -213,7 +213,7 @@ module internal InterfaceStubGenerator =
         let nm = 
             match arg.Name with 
             | None ->
-                if arg.Type.HasTypeDefinition && arg.Type.TypeDefinition.XmlDocSig = "T:Microsoft.FSharp.Core.unit" then "()" 
+                if arg.Type.HasTypeDefinition && arg.Type.TypeDefinition.CompiledName = "Microsoft.FSharp.Core.Unit" then "()"
```
Comment (dsyme) [inline src/fsharp/service/ServiceInterfaceStubGenerator.fs]: Is `Compiled` name really fully qualified at this point?  I thought that would just be `FSharpUnit`?

**Comment by dsyme** [comment]
@AviAvni Yes, stats please, thanks

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs (hunk 2 above)
Comment (AviAvni) [reply src/fsharp/service/ServiceInterfaceStubGenerator.fs]: just able to debug it and it's look like it's the CompiledName of unit is unit
image

**Comment by AviAvni** [comment]
for the custom ource the result are
```
~~~ Entity statistics: WithOptData 290, WithoutOptData 5272
~~~ Entity statistics: entity_compiled_name: 48
~~~ Entity statistics: entity_other_range: 0
~~~ Entity statistics: entity_kind: 106
~~~ Entity statistics: entity_xmldoc: 0
~~~ Entity statistics: entity_xmldocsig: 0
~~~ Entity statistics: entity_tycon_abbrev: 178
~~~ Entity statistics: entity_tycon_repr_accessibility: 46
~~~ Entity statistics: entity_accessiblity: 46
~~~ Entity statistics: entity_exn_info: 2
```

for the FSharp.Compiler.Private.fsproj
```
~~~ Entity statistics: WithOptData 7397, WithoutOptData 5978
~~~ Entity statistics: entity_compiled_name: 48
~~~ Entity statistics: entity_other_range: 3896
~~~ Entity statistics: entity_kind: 106
~~~ Entity statistics: entity_xmldoc: 1848
~~~ Entity statistics: entity_xmldocsig: 1839
~~~ Entity statistics: entity_tycon_abbrev: 737
~~~ Entity statistics: entity_tycon_repr_accessibility: 3253
~~~ Entity statistics: entity_accessiblity: 3253
~~~ Entity statistics: entity_exn_info: 502
```

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs [hunk 3]
```diff
@@ -213,7 +213,7 @@ module internal InterfaceStubGenerator =
         let nm = 
             match arg.Name with 
             | None ->
-                if arg.Type.HasTypeDefinition && arg.Type.TypeDefinition.CompiledName = "Microsoft.FSharp.Core.Unit" then "()" 
+                if arg.Type.HasTypeDefinition && arg.Type.TypeDefinition.CompiledName = "unit" then "()"
```
Comment (dsyme) [inline src/fsharp/service/ServiceInterfaceStubGenerator.fs]: Looks good, but can you check the namespace too, or something more?  Other people may define a type called "unit"

**Comment by dsyme** [comment]
OK, I think this is ready to go, with the exception of one minor thing in that "unit" comparison.

**Comment by dsyme** [comment]
Looks like a total reduction of about 35% of EntityData for referenced types.  

According to old statistics https://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html#analyzing-compiler-memory-usage EntityData is about 12% of long-lived memory usage, which would mean a 5% reduction overall for applications referencing lots of libraries, and maybe 2-3% for large applications like the compiler referencing relatively few libraries. Hopefully that's right!

**Comment by forki** [comment]
that sounds awesome!

**Comment by AviAvni** [comment]
@dsyme Done thanks 
I saw that I can apply the same technique for Typar and there is also a comment about it
or you have something else that have more priority

**Comment by dsyme** [comment]
> I saw that I can apply the same technique for Typar and there is also a comment about it
or you have something else that have more priority

I don't think it's worth it, Typar doesn't appear on the memory list

The next big memory hog for editing scenarios are the ``byte[]`` underlying [ByteFile](https://github.com/Microsoft/visualfsharp/blob/master/src/absil/ilread.fs#L222).  These are the in-memory read of referenced .NET binaries.  We use an in-memory `byte[]` to make collecting them (much) easier.

The question is what to do about this.  In truth I don't know....  Clearly we are using *way* too much memory for these and this really affects overall memory-usage for editing tools.  Also, in many cases, the byte[] are not substantially read from after some period of time (most the data gets transformed into the Abstract IL data structures, though this is done on-demand).  References to the ByteFile are still active, but the underlying memory is often not needed since we've already read most things from it.  And when a project is not active then the `byte[]` won't be read at all....  

I actually don't know what to do about this. Some notes:

* We already use an AgedLookup to ensure we don't create two ByteFile (actually ILBinaryReader) for the same file/time-stamp, caching a number of them.  So that's ok. 

* Could we release the `byte[]` on a "temporary basis" (e.g. if no access has been made to the ByteFile for 10 seconds) and re-read it later if needed?   But we would only want to do this if the underlying binary is not going to change.... That's quite hard to determine of course but could in theory be assumed for anything under Program Files.  But more and more binaries are under `packages\...` and these can change as packages are updated

* Another approach would be to avoid reading the IL sections of the binaries.  But in most cases we tend to reference "ref" assemblies that don't have this section in any case.  So I don't think that's a big win.

* Memory-mapping doesn't really help, since it chews through address space and also requires knowing that the binaries are not going to change... (see above).

* It's possible that seemingly bizarre suggestions like compressing the `byte[]` after it hasn't been used for 10 seconds would actually work very well.

I'll add these notes to the tracking issue

**Comment by auduchinok** [comment]
> Another approach would be to avoid reading the IL sections of the binaries. But in most cases we tend to reference "ref" assemblies that don't have this section in any case. So I don't think that's a big win.

It's not the case when referencing lots of assemblies with lots of symbols. Referencing ReSharper.SDK adds a lot to memory when extended completion is enabled. Not reading the IL would help much here.

> Memory-mapping doesn't really help, since it chews through address space and also requires knowing that the binaries are not going to change... (see above).

@dsyme It could be possible to read types with metadata and use file system watcher for these assemblies and invalidate imports when an assembly is changed.

**Comment by matthid** [comment]
@dsyme Just one note on this:

>  But more and more binaries are under packages\... and these can change as packages are updated

If with `packages/` you mean `C:\Users\<user>\.nuget\packages\microsoft.aspnetcore\2.0.1`

You can safely assume that this is immutable. In theory people can change stuff but both NuGet and Paket will yield undefined behavior if you mess with those files.

NuGet is designed to be immutable after a version has been published.

/cc @forki just to be sure
(Paket supports stuff like "local" packages, but I'd rather break that and have a faster/better compiler or have some other way of notification - In the end if you use the same packageId with different contents nobody can help you and you either know what you are doing or doing it severely wrong).

**Comment by dsyme** [comment]
> It's not the case when referencing lots of assemblies with lots of symbols. For example, referencing ReSharper.SDK adds a lot to memory when extended completion is enabled. Probably not reading the IL would help much here.

Yes, agreed that only reading metadata would help in these cases.  And it's probably really common in practice.

> @dsyme It could be possible to read types with metadata and use file system watcher for these assemblies and invalidate imports when an assembly is changed.

I think we effectively do that by abandoning the builder or going back to the Import-assemblies-step and dropping the old ILBinaryReader's and asociated objects on the floor.

But at the level of ByteFile the problem is that we never know when it's safe/right to drop the data....

**Comment by dsyme** [comment]
> You can safely assume that `.nuget\packages\...` is immutable. In theory people can change stuff but both NuGet and Paket will yield undefined behavior if you mess with those files.

Yes, agreed with that.

**Comment by vbfox** [comment]
From my measurements `ByteFile` is the single thing that make VS OOM for our team.

Both the compiler directly and via type providers maintain them in memory, and having a dll with some bitmap resources of a few dozen MB can trash VS memory very fast.

And from what I saw in the code only the tables (types, methods, ...) are used and not the IL or the resources or any other.

**Comment by auduchinok** [comment]
> Do you have a particular solution that we could use for testing?

[JetBrains/fsharp-support](https://github.com/JetBrains/fsharp-support) has many SDK symbol uses in `FSharp.Common` and `FSharp.Psi.Features` projects.

> Just add a package reference to ReSharper.SDK? We should add a sample solution that illustrates high ByteFile utilization in tests\projects\...

For a testing purpose we may add a simple project referencing the SDK, yes.
Any of [JetBrains.Rider.SDK](https://www.nuget.org/packages/JetBrains.Rider.SDK/2017.3.0) and [JetBrains.ReSharper.SDK](https://www.nuget.org/packages/JetBrains.ReSharper.SDK/) can be used (as they have most of ReSharper code in common).

**Comment by dsyme** [comment]
Great! On to ByteFile!!

**Pull request merged**